syntax = "proto3";

service ShapeService {
    rpc CreateShape(ShapeType) returns (CreateShapeResponse) {}
    rpc GetShape(ShapeId) returns (GetShapeResponse) {}
    rpc GetPerimetersGreaterThan(MinPerimeter) returns (stream GetPerimetersGreaterThanResponse) {}
    rpc GetTotalArea(stream ShapeId) returns (GetTotalAreaResponse) {}
    rpc GetAreas(stream ShapeId) returns (stream GetAreasResponse) {}
}

message CreateShapeResponse {
    Code status_code = 1;
    string message = 2;
    optional Shape shape = 3;
}

message GetShapeResponse {
    Code status_code = 1;
    string message = 2;
    optional Shape shape = 3;
}

message GetPerimetersGreaterThanResponse {
    Code status_code = 1;
    string message = 2;
    optional double perimeter = 3;
    optional Shape shape = 4;
}

message GetTotalAreaResponse {
    Code status_code = 1;
    string message = 2;
    optional double total_area = 3;
    repeated ShapeId valid_ids = 4;
    repeated ShapeId invalid_ids = 5;
}

message GetAreasResponse {
    Code status_code = 1;
    string message = 2;
    optional double area = 3;
    optional Shape shape = 4;
}

message MinPerimeter {
    double min_perimeter = 1;
}

message ShapeType {
    string shape_type = 1;
}

message ShapeId {
    string shape_id = 1;
}

message Shape {
    string shape_id = 1;
    string shape_type = 2;
    repeated ShapeCoord coords = 3;
}

message ShapeCoord {
    optional int32 x = 1;
    optional int32 y = 2;
}

// Start Custom Status Codes at 100 to not conflict with predefined gRPC status codes
// Enum MUST start at 0 for proto to compile
enum Code {
    OK = 0;
    INVALID_SHAPE = 100;
    INVALID_PERIMETER = 101;
    INVALID_SHAPE_ID = 102;
    SHAPE_NOT_FOUND = 103;
    AREA_NOT_FOUND = 104;
}